### PR TITLE
Shared loading and running pool

### PR DESCRIPTION
**Pull request by jakemac53**
I will add a test once getting agreement on a fix here. 

This should resolve https://github.com/dart-lang/test/issues/1159

### CODE REVIEW

File: pkgs/test_core/lib/src/runner/engine.dart [hunk 1]
```diff
@@ -220,7 +213,6 @@ class Engine {
   /// four times [concurrency].
   Engine({int concurrency, int maxSuites, String coverage})
```
Comment (natebosch) [inline pkgs/test_core/lib/src/runner/engine.dart]: We should remove the `maxSuites` argument since it has no meaning anymore. Was that used anywhere?

Comment (jakemac53) [reply pkgs/test_core/lib/src/runner/engine.dart]: It wasn't used within any of the `test` packages.

It is possible it was used elsewhere but it is somewhat difficult to tell for sure.

We are in a breaking release anyways so I will go ahead and just delete it.

**Comment by jakemac53** [comment]
PTAL - added a test

**Comment by jakemac53** [comment]
Note that bolt was using the `maxSuites` parameter but other than that internal is good to go. Removing that parameter for bolt gives the same behavior now (they also pass `concurrency: 1`) so we can land the change to remove that with the roll.

File: pkgs/test/test/runner/engine_test.dart [hunk 2]
```diff
@@ -275,4 +276,56 @@ void main() {
...
+          if (isLoadSuite) {
+            testsLoaded++;
+            maxLoadConcurrency = max(maxLoadConcurrency, testsLoaded);
+            expect(testsLoaded, lessThanOrEqualTo(concurrency));
+          } else {
+            testsRunning++;
+            maxTestConcurrency = max(maxTestConcurrency, testsRunning);
+            expect(testsRunning, lessThanOrEqualTo(concurrency));
+          }
+          await Future.delayed(Duration(milliseconds: 100));
```
Comment (grouma) [inline pkgs/test/test/runner/engine_test.dart]: Comment why this is necessary

File: pkgs/test_core/lib/src/runner/engine.dart [hunk 3]
```diff
@@ -265,35 +256,32 @@ class Engine {
...
-
+        var resource = await _runPool.request();
         LiveSuiteController controller;
-        if (suite is LoadSuite) {
-          await _onUnpaused;
-          controller = await _addLoadSuite(suite);
-          if (controller == null) {
-            loadResource.release();
-            return;
+        try {
```
Comment (grouma) [inline pkgs/test_core/lib/src/runner/engine.dart]: Why do we now need a try catch clause?

Comment (jakemac53) [reply pkgs/test_core/lib/src/runner/engine.dart]: This is just for the finally to ensure that we release the resource from the pool

Comment (jakemac53) [reply pkgs/test_core/lib/src/runner/engine.dart]: (we return a couple places from this function, and want to make sure we release it even if there is an error as well)

File: pkgs/test/test/runner/engine_test.dart (hunk 2 above)
Comment (jakemac53) [reply pkgs/test/test/runner/engine_test.dart]: done

**Pull request merged**